### PR TITLE
Make NIO compile on iOS - wrap sendfile() in #if's that match the import

### PR DESCRIPTION
**Pull request by helje5**
### Motivation:

NIO doesn't compile on iOS. Which I'd like to do.

### Modifications:

The `#if`'s used to protect the syscall do not match the `#if`'s used to `import`. Specifically the Darwin branch only applied to macOS, not watchOS nor homePodOS.

### Result:

NIO compiles on iOS. Thanks to NIO iPhone's can serve 100k RpS and will replace the IBM servers in the datacenter.

### CODE REVIEW

File: Sources/NIO/System.swift [hunk 1]
```diff
@@ -377,6 +377,8 @@ internal enum Posix {
                         written = 0
                     }
                     return result
+                #else
+                    let badOS = { fatalError("unsupported OS") }()
```
Comment (Lukasa) [inline Sources/NIO/System.swift]: Do we need the self-invoking function? I think just the `fatalError` itself would be fine.

Comment (helje5) [reply Sources/NIO/System.swift]: Yeah, right, just copied what the top-level was doing.

**Comment by helje5** [comment]
I wouldn't be unhappy if I could get a 1.2.2 w/ that as I have a small demo in the queue, which relies on this 😬

**Comment by Lukasa** [comment]
When's the demo? Just want to see what the scheduling is, because we dropped a semver minor into the codebase, so I'd like to try to smooth that out as much as possible, but it's a small one so we can probably bear going to 1.3 without too much pain if your demo is `O(very soon)`.

**Comment by Lukasa** [comment]
@swift-nio-bot test this please

**Comment by helje5** [comment]
I originally planned to release something just for fun today, not expecting this issue. But it isn't really important either. A 1.3.0 just sounds like "next release in two weeks".
I guess I can also just target a specific commit once you merge this.

**Comment by Lukasa** [comment]
Sure thing, so I'd tentatively be shooting to release 1.3.0 middle of next week in the absence of anything else. Would that work for you?

**Comment by helje5** [comment]
Sure, sounds good to me.

File: Sources/NIO/System.swift [hunk 2]
```diff
@@ -363,12 +363,12 @@ internal enum Posix {
...
+                #if os(macOS) || os(iOS) || os(watchOS) || os(tvOS)
                     var w: off_t = off_t(count)
                     let result: CInt = Darwin.sendfile(fd, descriptor, offset, &w, nil, 0)
                     written = w
                     return ssize_t(result)
-                #else
+                #elseif os(Linux) || os(FreeBSD) || os(Android)
                     var off: off_t = offset
                     let result: ssize_t = Glibc.sendfile(descriptor, fd, &off, count)
                     if result >= 0 {
```
Comment (weissi) [inline Sources/NIO/System.swift]: FreeBSD’s sendfile is more like Darwin’s but doesn’t seem to be quite it either. Definitely not like the Linux one though. It should fall into the unsupported platform for now.

https://www.freebsd.org/cgi/man.cgi?query=sendfile&sektion=2

**Comment by helje5** [comment]
Well, I didn't add FreeBSD, so maybe that is another patch? (presumably remove it everywhere because Swift on FreeBSD doesn't work anyways, right?)
Also, for unsupported platforms, maybe sendfile should rather throw? An upper layer could switch to a different delivery method, it is not really a fatal error. But again, out of scope for this specific patch.

File: Sources/NIO/System.swift (hunk 2 above)
Comment (normanmaurer) [reply Sources/NIO/System.swift]: @weissi I think @helje5 is right and we can do this as a follow up. WDYT ?

Comment (weissi) [reply Sources/NIO/System.swift]: Sure 👍👍

Comment (normanmaurer) [reply Sources/NIO/System.swift]: ok let me merge then :)

**Comment by normanmaurer** [comment]
Merged as two commits (94718bcaa86d244412849c931a22a65211fc175c and b06b71208595578ceb2304a051e01adc3330944e)

@helje5 thanks!

**Pull request closed without merging**